metaflow/cmd/code/__init__.py

```diff
@@ -0,0 +1,229 @@
+                if not output:
+                    echo(f"❗  {target_file} not in the target directory, skipping")
+    return diffs if output else None
+
+
+def run_op(runspec: str, op: Callable[..., None], op_args: Dict[str, Any]) -> None:
+    tmp = None
+    try:
+        tmp = extract_code_package(runspec, EXCLUSIONS)
+        op(tmp.name, **op_args)
+    finally:
+        if tmp and os.path.exists(tmp.name):
+            shutil.rmtree(tmp.name)
+
+
+def run_op_diff_runs(source_run: str, target_run: str) -> None:
+    source_tmp = None
+    target_tmp = None
+    try:
+        source_tmp = extract_code_package(source_run, EXCLUSIONS)
+        target_tmp = extract_code_package(target_run, EXCLUSIONS)
+        perform_diff(source_tmp.name, target_tmp.name)
+    finally:
+        for d in [source_tmp, target_tmp]:
+            if d and os.path.exists(d.name):
+                shutil.rmtree(d.name)
+
+
+def op_diff(tmpdir: str) -> None:
+    perform_diff(tmpdir)
+
+
+def op_pull(tmpdir: str, dst: str) -> None:
+    if os.path.exists(dst):
+        echo(f"❌  Directory *{dst}* already exists")
+    else:
+        shutil.move(tmpdir, dst)
+        echo(f"Code downloaded to *{dst}*")
+
+
+def op_patch(tmpdir: str, dst: str) -> None:
+    diffs = perform_diff(tmpdir, output=True) or []
+    with open(dst, "w", encoding="utf-8") as f:
+        for out in diffs:
+            out = out.replace(tmpdir, "/.")
+            out = out.replace("+++ b/./", "+++ b/")
+            out = out.replace("--- b/./", "--- b/")
+            out = out.replace("--- a/./", "--- a/")
+            out = out.replace("+++ a/./", "+++ a/")
+            f.write(out)
+    echo(f"Patch saved in *{dst}*")
+    path = run(
+        ["git", "rev-parse", "--show-prefix"], text=True, stdout=PIPE
+    ).stdout.strip()
+    if path:
+        diropt = f" --directory={path.rstrip('/')}"
+    else:
+        diropt = ""
+    echo("Apply the patch by running:")
+    echo_always(
+        f"git apply --verbose{diropt} {dst}", highlight=True, bold=True, err=True
+    )
+
+
+@code.command()
+@click.argument("metaflow_run")
+def diff(metaflow_run: str) -> None:
+    """
+    Do a 'git diff' of the current directory and a Metaflow run.
+    """
+    run_op(metaflow_run, op_diff, {})
+
+
+@code.command()
+@click.argument("source_run")
+@click.argument("target_run")
+def diff_runs(source_run: str, target_run: str) -> None:
+    """
+    Do a 'git diff' between two Metaflow runs.
+    """
+    run_op_diff_runs(source_run, target_run)
+
+
+@code.command()
+@click.argument("metaflow_run")
+@click.option(
+    "--dir", help="Destination directory (default: {runspec}_code)", default=None
+)
+def pull(metaflow_run: str, dir: Optional[str] = None) -> None:
+    """
+    Pull the code of a Metaflow run.
+    """
+    if dir is None:
+        dir = metaflow_run.lower().replace("/", "_") + "_code"
+    run_op(metaflow_run, op_pull, {"dst": dir})
+
+
+@code.command()
+@click.argument("metaflow_run")
```

Comment:
Nit: `metaflow-run` since most of our CLI args typically have `-` and not `_`'s.